### DISCUSSION TITLE
TabItem State Styling requires internal knowledge of control's coding (Bad Design Karma!!!)

### DISCUSSION THREAD

**Discussion by KentSwan** [AvaloniaUI/Avalonia#14131] [General]
## Is your feature request related to a problem? Please describe.

To style the content of a tabitem I have the following  style settings.

		<Style Selector="TabControl">
			<Setter Property="Padding" Value="0 0 0 0"/>
			<Setter Property="Margin" Value="0 0 0 0"/>
		</Style>
		<Style Selector="TabItem">
			<Setter Property="BorderBrush" Value="Blue"/>
			<Setter Property="BorderThickness" Value="2"/>
			<Setter Property="CornerRadius" Value="8 8 0 0"/>
			<Setter Property="Background" Value="White"/>
			<Setter Property="Foreground" Value="Black"/>
		</Style>
		<Style Selector="TabItem:selected /template/ ContentPresenter#PART_ContentPresenter">
			<Setter Property="Background" Value="Black"/>
			<Setter Property="Foreground" Value="Yellow"/>
			<Setter Property="FontWeight" Value="Bold"/>
			<Setter Property="FontStyle" Value="Italic"/>
		</Style>

The first TabItem style sets the unselected tab style elements
The second Tabitem style set the selected tab style elements.

My issue is that the :selected attribute tag should be suffient to do this without having to know and specify control internal information.

## Describe the solution you'd like

When you add state attributes to a control component for styling, the backing concept is that the properties set override the base properties of that control component based on the state attribute being true or ignored if false.

I would thus like to have 
           <Style Selector="TabItem:selected /template/ ContentPresenter#PART_ContentPresenter">
and
           <Style Selector="TabItem:selected">
be equivalent.

I contend that the two :selector attribute examples should be equivilent and sufficient to map the setter to the control components' internal template,   For long-term regression consistency, all state :attributes should always have a default template property mappings along with an ability to override that mapping by being explicit.  

This conforms to the concept that normally implementation-specific internal knowledge, reference or access should not be required for ordinary use but an override argument syntax is provided when it may be required (typically this would be when using the control as a component part of a more complex control.  In the above example case,  the two forms of the selectors should be functionally identical where the additional mapping, in this case, would be the default internal mapping.

## Describe alternatives you've considered

None as yet. It's irritating to have to go into the control code to obtain the mappings required to get to the state styling properties overloads.

## Additional context
None.

**Comment by jp2masa** [comment]
This has been discussed several times, and it's by design, e.g. https://github.com/AvaloniaUI/Avalonia/pull/2662#issuecomment-515764732.

**Comment by KentSwan** [comment]
So we have to go rooting around in every control to locate the templates that are associated with state-dependent style properties so we can set them?   If this is true then can I assume that "/template/ ContentPresenter#PART_ContentPresenter" will always work in every use case? If not then where is the how-to documentation other than in the rtfc ("Read the F***ing Code.") 

As an aside, I've been doing software design and implementation for more years than I can count and that includes a ton of API's and their implementations.  My comment on that "It's by design" doesn't mean that the design choices are correct or follow accepted object interface practices and their underlying implementations.

Sorry about the soapbox but you hit one of my pet peeves with that "it is what it is" statement.

**Comment by jp2masa** [comment]
Personally, I don't like this design choice either, because it doesn't make sense to set a local value and then see the value changing due to a style. I don't make decisions, though, and the core team has to consider many aspects when deciding. In this case, it was related to WPF compatibility IIRC.

**Comment by thevortexcloud** [comment]
>  every control to locate the templates that are associated

Yes. They are "lookless" controls. Which means they can look like anything. Most people just happen to be using the official themes, but there is nothing stopping somebody else from making it look and visually function completely differently. The only thing control templates have to declare are `PART_` parts (and even then, nothing can force someone to actually implement them. It just often leads to a control not functioning properly). Which you can find by looking at the C# code for the control. EG this is the declaration for the slider: 

https://github.com/AvaloniaUI/Avalonia/blob/38e839997d2c204548e6fad396c178780a010cb1/src/Avalonia.Controls/Slider.cs#L43-L50

But the parts can still be located really anywhere in the control. A lot of controls are also inherited so you sometimes need to go digging pretty deep to find the full list of parts.

**Comment by thevortexcloud** [comment]
> it was related to WPF compatibility IIRC.

This indeed fairly similar to how WPF works and it's also just as annoying trying to find a list of all the control parts.

**Comment by KentSwan** [comment]
Having state-dependent styles is a consistent way to implement visual animation that tells the user what's going on. The issue for an application designer is that they need the styling of things to be relatively consistent throughout the application which means that those design choices may not be what the API defaults to.  In any case, the list of available state : attributes are the same across the board for every control that implements them (which is most of them).  The application UI coder should NEVER have to know anything about how an internal aspect of how a control is coded or a subcomponent named because the API and Control coders can arbitrarily change those (most frequently because of a required architectural reorganization) which would then break ALL dependent user applications.  

Again, this means that this internal exposed mapping requirement, while currently necessary, is bogus.  

Axaml is already compiled and translated to map both the view and the data context by interpreting the source syntax and mapping it to the implementation.  With the advent of code generators, the API and control maintainers should be able to craft a consistent way of extending the compilation to internally map or accept a template override to the underlying code.  Granted the knowledge required is likely massive but those with experience should be able to make it happen.  

If my suggestion is accepted then future axaml users would be able to code without fussing but having the option to add a different mapping would be regression compliant and thus not break the existing code base.

Question: Does anyone in the Avalonia development team that have design knowledge and implementation control look at the issues we bring up?

**Comment by thevortexcloud** [comment]
> a consistent way of extending the compilation to internally map or accept a template override to the underlying code

XAML can be loaded at runtime, not just at compile time. Control templates can also be defined at runtime. That said I would like it to be possible for compile errors to be raised if a `PART_` is not defined for a template in XAML. It would make making/maintaining control templates far less painful.

>  would be regression compliant and thus not break the existing code base.

I fail to see how this is possible. If you are reliant on a theme working in a specific way and then they suddenly change what a pseudo selector does then you now have a breaking change. You are better off asking for the theme author to make their templates use template bindings when possible. `TabItem` for example should already do this for the border control on the fluent theme

https://github.com/AvaloniaUI/Avalonia/blob/38e839997d2c204548e6fad396c178780a010cb1/src/Avalonia.Themes.Fluent/Controls/TabItem.xaml#L30-L36

In which case setting the `Background` property on the `TabItem` would change the background of that `Border` control. However this is still reliant on having a well designed control template, and it wont work for every property you may want to change. Which is why the `/template/` selector exists.

**Comment by KentSwan** [comment]
re: TheVortexCloud:

I could live with that IF all control and theme designers had a consistent way of causing that information to be exported as readable usable documentation.  RTFC is not a practical Avalonia API option for anyone except the API and Control maintainers. I do have to concede that you do have a valid point for complex composited controls and the attendant WPF baggage.  

In my way of thinking, state attributes are boolean elements that apply to alternate values for a specific defined sub-collection of properties of the control or control subcomponent.  The binding of the state to the internal PART_ should also be internal so that using the state attribute in your application UI design is independent (largely) of the internal naming choices.  Again this is good API implementation practice.

**Reply by thevortexcloud** [reply]
> . RTFC is not a practical Avalonia API option for anyone except the API and Control maintainers.

I imagine that information could probably be auto generated based on the attributes in the control class and then dumped onto the docs site somewhere so people can at least see it exists. The only control I can see there that has it at all is the colour picker

https://github.com/AvaloniaUI/avalonia-docs/blob/178e1e966887c8f1c57ad9ecbbd406be19205a14/docs/reference/controls/colorpicker/colorview.md?plain=1#L50-L55

But we would probably need someone who is more familiar with the docs site build system to say if that is feasible. That said, I agree that reading the code is not a great option.

>  Again this is good API implementation practice.

Unfortunately it's far too late to change this. It's pretty much the system Microsoft settled on [many years ago with WPF.](https://learn.microsoft.com/en-us/dotnet/api/system.windows.templatepartattribute?view=windowsdesktop-8.0&redirectedfrom=MSDN) Although in WPF it's far messier since they don't have the concept of psuedo selectors at all, so it's all hacked together with triggers and single frame animations.

**Comment by KentSwan** [comment]
I did a little deeper diving and pulled the TabItem.xaml file to examine how it exposed things and the issue with having to specify the internal template.  In actual fact, you shouldn't have to because all of the information to bind the pseudo selectors to the internal template designator is built into the code.  See the annotated jpg image below.  As you can see the internal binding of the pseudo selectors point back to the block of properties that are exposed via template binding.  I looked at several other control axaml files and they follow the same pattern but sometimes with a different PART name.  If the pseudo selector processing (which is an Avalonia ONLY feature) picked this up as the default (unless overridden by the user) then it, automatically, would accomplish the required decoupling that I've pointed out. 

![TabItem xaml_Annotated](https://github.com/AvaloniaUI/Avalonia/assets/7853729/0c03c8ea-aa56-4015-a1e5-c8d249e167fb)

**Reply by timunie** [reply]
Would a link to the original styles in the docs improve the situation for you? I'm just thinking out loud here. That we change the API is unlikely.

**Reply by thevortexcloud** [reply]
For me, a list of all the control parts would be nice.

**Reply by timunie** [reply]
Probably, yes. We have an attribute for it now, so may be possible to extract 🤔

**Reply by thevortexcloud** [reply]
That was my exact thought.

> I imagine that information could probably be auto generated based on the attributes in the control class and then dumped onto the docs site somewhere so people can at least see it exists. The only control I can see there that has it at all is the colour picker

Obviously it would be nice if it was automatically kept in sync with what ever the latest release version is. I am however not sure how the doc site/doc build system works. In any case, it should be fairly easy to generate a bunch of markup for all controls in a specific namespace based on the attribute values. I am guessing some github action could possibly do that. But it would likely mean adding a bunch of C# code to the build system.

**Reply by timunie** [reply]
docs is written by hand, but we are trying to get the API refs back to life again.

**Reply by thevortexcloud** [reply]
I figured it was completely hand written. That said, some sort of quick script to dump the attribute values and then manually merged into the control page would still be better than nothing.

**Reply by thevortexcloud** [reply]
Oh also runtime/compile exceptions would also be great for a templated control that does not have a part defined in the theme. Right now making custom templates is a pain since they can break for no obvious reason if Avalonia decides to change some of the names around/suddenly requires different parts.

**Reply by timunie** [reply]
> Oh also runtime/compile exceptions would also be great for a templated control that does not have a part defined in the theme. Right now making custom templates is a pain since they can break for no obvious reason if Avalonia decides to change some of the names around/suddenly requires different parts.

That should be it's own feature request I guess.

**Reply by thevortexcloud** [reply]
I'll probably write one up then when I get time.

**Comment by KentSwan** [comment]
This brings up an interesting question.  Is there an addendum info markup syntax for both C# and axaml/xaml that allows the API design and support teams to incorporate additional embedded textual documentation about API things directly within the deployed files in a way that would be transparent to both the design-time and run-time compilers?  A convention using this would allow the up-to-date information for the user interfaces to be continuously extracted from those files into the API docs we need to use them.

**Comment by KentSwan** [comment]
This is another valid reason to use proper API design rules and NOT have to use exposed internal control design elements.  The API for for the use of a control should NEVER require that you reference internal design information. PERIOD! FULL STOP.  

Not implementing an API methodology that allows designers to follow this rule will eventually cause the API design to stagnate and eventually the API will fail the user community.  The is because the API cannot evolve at some point, just because you can't change the internal architecture of a control without replacing it in its entirety with a newly named control unless you don't care about breaking older uses of that control. 

In this regard, I'm suggesting that, since axaml pseudo-selectors are Avalonia-only constructs, their implementation should be evolved in a way that makes the internal reference mapping INTERNAL while allowing the existing syntax to remain. As controls are updated to the new internal reference form the unwanted external reference becomes an ignorable artifact which should be indicated by a "deprecation compile warning".

**Reply by MikeCodesDotNET** [reply]
We value your input on the API design but request a more respectful tone in your communication. Your point is noted, but it's crucial to maintain a constructive environment for discussion.

**Comment by KentSwan** [comment]
I truly apologize for the tone and will strive not to repeat it in the future.  

Like you all, I've been designing system-level software for a ton of years along with too many API's to count. Unfortunately, the partitioning and exposure aspects of design are one of my hot buttons.  Over those many years, I've learned, the hard way, that violating good design practices such as I pointed out may well get you through the here and now but will eventually come back, much more expensively, to haunt you.  The downstream result is often that the project will die by going into an unrevisable historical stasis and will eventually be abandoned by it's users because it cannot move forward.  Again I humbly apologize for the tone but not the underlying message.
